### ISSUE TITLE
execution of code in quarto cell sometimes works, sometimes does not

### ISSUE THREAD

**Issue by rmflight** [posit-dev/positron#3945]
## Positron Version: 2024.07.0 build 21
Code - OSS Version: 1.90.0
Commit: 82b335e27545e804b7fd806db00477a6eb49951d
Date: 2024-07-09T03:06:22.884Z
Electron: 29.4.0
Chromium: 122.0.6261.156
Node.js: 20.9.0
V8: 12.2.281.27-electron.0
OS: Linux x64 6.9.3-76060903-generic
R: 4.3.0

## Steps to reproduce the issue:

1. Create a new quarto document
2. Insert a code cell
3. Try to evaluate code by line or piped statement
4. Only partial evaluation happens

[Screencast from 07-09-2024 04:13:08 PM.webm](https://github.com/posit-dev/positron/assets/1509626/03dd5510-86b7-44f2-9223-7ca2bb4eadc8)

In this example, I tried to run the code block containing

```
data(mtcars)
```

using `Ctrl+Enter`. But only the `data` part is called. Another code block, with a `library` call, seemed to work fine. Also, executing the full block, via `Ctrl+Shift+Enter` also works (not shown in recording).

I've also had this fail for using cells containing code that spans multiple lines with pipes, and only a single or a couple of lines of the pipe will be executed.

## What did you expect to happen?

Expect full code statement to be evaluated.

## Were there any error messages in the output or Developer Tools console?

No.

**Comment by jmcphers** [comment]
I couldn't reproduce this after typing out the example from the screenshot above.

<img width="1009" alt="image" src="https://github.com/posit-dev/positron/assets/470418/b7dcdcff-03b6-466a-aaf0-76e3b9f8575d">

@rmflight Could you share any other examples, maybe one with the partial pipes?

**Comment by rmflight** [comment]
Sure thing. I also checked the logs, and there is something weird in there.

Better example

````
---
title: "Untitled"
format: html
---

```{r}
#| label: test-block
data("mtcars")
```

```{r}
#| label: run-ggplot
library(ggplot2)
```

```{r}
#| label: test-piped
mtcars |>
    dplyr::filter(mpg > 20) |>
    dplyr::summarise(n = dplyr::n())
```

```{r}
#| label: another-piped
mtcars |>
    dplyr::filter(mpg < 20) |>
    dplyr::summarise(n = dplyr::n())
```
````

Screencast of today: 

[Screencast from 07-10-2024 03:28:48 PM.webm](https://github.com/posit-dev/positron/assets/1509626/01a9ae21-e798-422a-ab01-a908fcf9ac3c)

And log output. 

[log_2024-07-10-15-25.log](https://github.com/user-attachments/files/16168078/log_2024-07-10-15-25.log)

You can see that only one line of the pipe is being sent at a time, when I expect the full thing to be sent.

In addition, ark is restarting my R kernel when it tries to access the variables so I can see them in my dplyr call. 

Also, I'm running R 4.3.0, in case that matters.

**Comment by juliasilge** [comment]
Might this be related to #3792 as well, and something going wrong with the mapping between `.qmd` and `.R`?

**Comment by rmflight** [comment]
I'm not sure what has changed from build 21 to build 52, but this issue seems to have gone away as of now.

I installed latest release this morning, and it seems to be behaving properly in my `.qmd` documents now.

**Comment by coreysparks** [comment]
I am still seeing this issue with today's build, no rendering of output within a quarto document

`---
title: "Untitled"
format: html
---


```{r}
mean(rnorm(10))
hist(rnorm(100))

````

![image](https://github.com/user-attachments/assets/e8337287-887d-433a-861b-d2abc024fb7f)
should look like this from Rstudio

![image](https://github.com/user-attachments/assets/ccf8bef7-1cf7-4436-9a44-81b5fd9c1cb5)

**Comment by juliasilge** [comment]
Ah @coreysparks I believe you are expecting to see _inline_ output for a Quarto document, which is not what this issue is about. You can read more about using Quarto in an inline mode (vs. a "send to console" mode) in #1418, #1533, #3718, and #4033.

**Comment by rmflight** [comment]
Positron Version: 2024.08.0 build 24

I keep downloading the new releases to see if this has been fixed yet, and to see if I can figure out *where* exactly the issue comes from.

I just noticed today, that if I open a project with a quarto document in it, I can pass any lines already written in the document when I opened the project, but if I try and write any **new** r code blocks, or start a new document, it doesn't work.

Nothing shown in the output log besides R starting up.

**Comment by DavisVaughan** [comment]
I'm seeing this in the logs that @rmflight posted, making me think this is indeed related to #3792 

```
[R]   2024-07-10T19:29:37.116795Z ERROR  Requesting line 27 but only 17 lines exist. Document: '#
[R] #
[R] #
[R] #
[R] #
[R] #
[R] #| label: test-block
[R] data("mtcars")
[R] #
[R] #
[R] #
[R] #| label: run-ggplot
[R] library(ggplot2)
[R] #
[R] #
[R] #
[R] '.
[R]     at crates/ark/src/lsp/encoding.rs:84
```

It would be nice if we saw a full traceback in the logs here, so we can determine where exactly this is coming from in ark.

So one thing we should definitely do is try and either convert this error log to an actual error, or maybe try and force a backtrace capture right here and log it (I _think_ we can do that), and then request that @rmflight reupload logs after triggering it again.

**Comment by rmflight** [comment]
I'm not sure if the `Positron Version: 2024.09.0 build 27` has what you wanted @DavisVaughan , or if I'm waiting on another build, but I went ahead an forced the error again.

Logs attached from today.

[log_2024-09-13-14-56.log](https://github.com/user-attachments/files/16998340/log_2024-09-13-14-56.log)

**Comment by DavisVaughan** [comment]
Thanks, looks like the backtrace itself didn't prove that useful (probably because you are on a release build with some debug symbols stripped out) but this is still useful for us to start playing with

```
[R]   2024-09-13T18:52:04.611438Z ERROR  Requesting line 33 but only 31 lines exist.
[R] 
[R] Document:
[R] #
[R] #
[R] #
[R] #
[R] #
[R] #
[R] #| label: test-block
[R] data("mtcars")
[R] #
[R] #
[R] #
[R] #| label: run-ggplot
[R] library(ggplot2)
[R] #
[R] #
[R] #
[R] #| label: test-piped
[R] mtcars |>
[R]     dplyr::filter(mpg > 20) |>
[R]     dplyr::summarise(n = dplyr::n())
[R] #
[R] #
[R] #
[R] #| label: another-piped
[R] mtcars |>
[R]     dplyr::filter(mpg < 20) |>
[R]     dplyr::summarise(n = dplyr::n())
[R] #
[R] #
[R] #
[R] 
[R] 
[R] 
[R]     at crates/ark/src/lsp/encoding.rs:86
```

In particular, reverse engineering the qmd...

``````

---
title: "Untitled"
format: html
---

```{r}
#| label: test-block
data("mtcars")
```

```{r}
#| label: run-ggplot
library(ggplot2)
```

```{r}
#| label: test-piped
mtcars |>
  dplyr::filter(mpg > 20) |>
  dplyr::summarise(n = dplyr::n())
```

```{r}
#| label: another-piped
mtcars |>
  dplyr::filter(mpg < 20) |>
  dplyr::summarise(n = dplyr::n())
```


``````

I still can't force the panic when playing around with this qmd but maybe someone on the team can

**Comment by rmflight** [comment]
It seems to evaluate pre-existing code blocks OK when I open the file. It's when I **add a new code block** that it seems to have trouble with.

Also, I have custom snippet for QMD files, that when I hit "r" -> "tab" I get the option to insert a qmd code block. 😉

**Comment by juliasilge** [comment]
We have a possible fix for you to try out @rmflight if you are willing, which looks like it has helped in #3792. Can you please try:

- Download the artifact from this GH action run: https://github.com/quarto-dev/quarto/actions/runs/11508346112
- Unzip the artifact
- Install the `.vsix` in Positron using the command "Extensions: Install from VSIX"
- Restart the extension host, as instructed by the button
- See what happens as you try to use Quarto???? 🤞

**Comment by rmflight** [comment]
@juliasilge I think it works! I just went through my normal testing file, adding more and more multiline blocks, and it just works, as I expect it to. And yeah, #3792 sounds suspiciously like almost the same issue, in that whatever is already written seems to work, but creating new blocks is when it decides to crash. I'm still curious why it seems almost impossible for all y'all to reproduce ....

I will attempt to use Positron on another document in an actual analysis project, and see how it goes as well.

**Comment by juliasilge** [comment]
Thank you so much for being willing to try it out @rmflight! We would really appreciate it if you could continue to test out this proposed fix. ❤

**Comment by juliasilge** [comment]
We've got another new dev version of the Quarto extension here with a cleaned-up version of the fix, if you would like to again download, unzip, and install the `.vsix`: https://github.com/quarto-dev/quarto/actions/runs/11581960140

We will do an official release of the Quarto extension with this fix probably before the end of the week.

Since we don't have an internally reproducible way of generating the original problem, I am going to move this issue straight to "Done" (no QA verification). I used the `.vsix` from main as of now to run through normal Quarto workflows, and all appears to be working well! Thank you again for your feedback and patience. 🙌
